### PR TITLE
feat: added tolerance for web3 and DB errors, improved logging for balance cache worker

### PR DESCRIPTION
**Pull request by alexkroeger**
# Description

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Test changes on the staging environment with [Matcha API Staging](https://api-staging.matcha.xyz).

    -   [ ] SRA/Limit orders
    -   [ ] Swap endpoints
    -   [ ] Meta transaction endpoints
    -   [ ] Depth charts

    For more information see `0x API Matcha smoke test runbook` in Quip.

### CODE REVIEW

File: src/runners/rfq_maker_balance_cache_runner.ts [hunk 1]
```diff
@@ -107,7 +113,21 @@ async function runRfqBalanceCacheAsync(
...
+        if (dbWriteErrors >= MAX_DB_WRITE_ERRORS) {
+            throw new Error(`Too many failures to write to DB (reached limit of ${MAX_DB_WRITE_ERRORS})`);
+        }
+        let newBlock: number;
+        try {
+            newBlock = await web3Wrapper.getBlockNumberAsync();
+        } catch (err) {
+            requestErrors += 1;
+            logger.error(err);
+            break;
```
Comment (PirosB3) [inline src/runners/rfq_maker_balance_cache_runner.ts:128]: Maybe `continue`?

File: src/runners/rfq_maker_balance_cache_runner.ts [hunk 2]
```diff
@@ -119,7 +139,12 @@ async function runRfqBalanceCacheAsync(
                 'Found new block',
             );
 
-            await cacheRfqBalancesAsync(connection, balanceCheckerContractInterface, true, workerId);
+            try {
+                await cacheRfqBalancesAsync(connection, balanceCheckerContractInterface, true, workerId);
+            } catch (err) {
+                logger.error(err);
+                break;
```
Comment (PirosB3) [inline src/runners/rfq_maker_balance_cache_runner.ts:146]: Maybe `continue`?

Comment (alexkroeger) [reply src/runners/rfq_maker_balance_cache_runner.ts:146]: 👍

File: src/runners/rfq_maker_balance_cache_runner.ts (hunk 1 above)
Comment (alexkroeger) [reply src/runners/rfq_maker_balance_cache_runner.ts:128]: 👍

File: src/runners/rfq_maker_balance_cache_runner.ts [hunk 3]
```diff
@@ -141,7 +166,13 @@ export async function cacheRfqBalancesAsync(
     const balancesCallInput = splitValues(makerTokens);
 
     const updateTime = new Date();
-    const erc20Balances = await getErc20BalancesAsync(balanceCheckerContractInterface, balancesCallInput, codeOverride);
+    let erc20Balances: string[];
+    try {
+        erc20Balances = await getErc20BalancesAsync(balanceCheckerContractInterface, balancesCallInput, codeOverride);
+    } catch (err) {
+        requestErrors += 1;
```
Comment (PirosB3) [inline src/runners/rfq_maker_balance_cache_runner.ts:173]: You are exporting this function, which means that it's usage can happen even outside of this file/module. However, this function does change state witin the module, which isn't the best design imo. Can you update the `requestErrors` outside of `cacheRfqBalancesAsync` instead?

**Comment by phil-ociraptor** [comment]
Thoughts on using a library such as this one? https://www.npmjs.com/package/async-retry

It's from Vercel, and has a lot of usage, which are good signs.

The usage of it would then be:

```
import retry from 'async-retry'

const newBlock = await retry(web3Wrapper.getBlockNumberAsync, { retries: 10 })
```

File: src/runners/rfq_maker_balance_cache_runner.ts (hunk 3 above)
Comment (alexkroeger) [reply src/runners/rfq_maker_balance_cache_runner.ts]: abstracted this away

**Review by PirosB3 — APPROVED** [review]

**Comment by alexkroeger** [comment]
deploy staging

**Pull request merged**